python/lsst/ts/observatory/control/auxtel/atcalsys.py

```diff
@@ -208,6 +213,81 @@ async def prepare_for_flat(self, config_name: str) -> None:
         )
 
+    async def calculate_optimized_exposure_times(
+        self, wavelengths: list, config_data: dict
+    ) -> list:
+        """Calculates the exposure times for the fiber spectrograph and
+        the electrometer given camera exposure times and other
+        relevant configuration paramaters.
+
+        This includes wavelength dependent information for the fiber
+        spectrograph and also includes the logic for the size of
+        the electrometer buffer.
+
+        Parameters
+        ----------
+        wavelengths : `list`
+            List of all wavelengths for this exposure list
+        config_data : `dict`
+            All information from configuration file
+
+        Returns
+        -------
+        exposure_list : `array`
+            List of exposure information, including wavelength and camera,
+            fiberspectrograph and electrometer exposure times.
+
+        """
+        exposures = []
+        for wavelength in wavelengths:
+            for exptime in config_data["exposure_times"]:
+                if config_data["use_electrometer"]:
+                    time_sep = (
+                        config_data["electrometer_integration_time"] * 3.07
+                    ) + 0.00254
+                    max_exp_time = 16667 * time_sep  # buffer size
+                    if exptime > max_exp_time:
+                        electrometer_exptime = max_exp_time
+                        self.log.info(
+                            f"Electrometer exposure time reduced to {max_exp_time}"
+                        )
+                    else:
+                        electrometer_exptime = exptime
+                else:
+                    electrometer_exptime = np.nan
+
+                if config_data["use_fiberspectrograph"]:
+                    base_exptime = 1  # sec
+                    entry_slit_multiplier = 6 / (config_data["entrance_slit"] + 1)
+                    exit_slit_multiplier = 6 / (config_data["exit_slit"] + 1)
+                    # TODO (DM-45235): Improve these multipliers with testing
+                    # Also, include some wavelength dependence based on testing
+                    fiberspectrograph_exptime = (
+                        base_exptime * entry_slit_multiplier * exit_slit_multiplier
+                    )
+                else:
+                    fiberspectrograph_exptime = np.nan
+
+                exposures.append(
+                    [
+                        wavelength,
+                        exptime,
+                        electrometer_exptime,
+                        fiberspectrograph_exptime,
+                    ]
+                )
+
+        exposure_list = np.array(
+            [np.vstack(exposures)],
+            dtype=[
+                ("wavelength", float),
+                ("camera", float),
+                ("fiberspectrograph", float),
+                ("electrometer", float),
+            ],
+        )
+        return exposure_list
```

Comment:
you are documenting the return type as a `list` but this is a `numpy` array no? Maybe this should be a list of dictionaries? Something like `list[dict[str, float]]`? then make each "exposure" be a dict with the elements you are now using the `dtype` for?

Reply:
I went back and forth on the best kind of data type to use here and finally landed on a numpy rec array. To me, it's a bit cleaner than a list of dictionaries.

Reply:
The numpy array is really useful for data analysis type applications, but for this one, not so much. If you think about the two alternatives, you are basically generating a list of wavelength, camera, fiber spec, electrometer exposure times, so a list of something. Whereas the numpy array is a table with different columns. I think the list of dictionary makes a lot more sense. 

Also, note, for instance, you can't provide a good type hint for  the data when it is a numpy array. With a list of dict, it is more clear. 

Also, if you want to take this one step further, you could make a data class to store this information and make the return type a list of this object instead. That would be even better. It would look like this:

```
@dataclass
class ATCalsysExposure:
    wavelength: float
    camera: float
    fiberspectrograph: float | None
    electrometer: float | None
```

and the return type would be `list[ ATCalsysExposure ]`.

Note that here I am also assuming you will use `None` for when there's no exposure of a particular type.